### PR TITLE
[VALIDATED_STATE] - Add state information to the `Decide` event

### PR DESCRIPTION
**Pull request by shenkeyao**
Closes https://github.com/EspressoSystems/HotShot/issues/2695.

### This PR: 
* Creates an application-specific `StateDelta` type and adds its test impl.
* Adds `StateDelta` to be an associated type of `ValidatedState`.
* Updates `ValidatedState::validate_and_apply_header` and `::genesis` to return the state delta as well.
* Adds the state and state delta to the leaf chain in the `Decide` event and the saved view.

### Key places to review: 
* `crates/example-types/src/state_types.rs`.

### CODE REVIEW

**Comment by shenkeyao** [comment]
I'm going to send an update to wrap the fields in `Event` into `Arc`.

File: crates/task-impls/src/consensus.rs [hunk 1]
```diff
@@ -827,6 +831,8 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                             event: EventType::Decide {
                                 leaf_chain: Arc::new(leaf_views),
                                 qc: Arc::new(new_decide_qc.unwrap()),
+                                validated_state: state,
+                                state_delta: Arc::new(state_delta),
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/task-impls/src/consensus.rs]: Sounds good! I've moved the state and delta to the leaf chain.

File: crates/types/src/traits/states.rs [hunk 2]
```diff
@@ -28,18 +31,20 @@ pub trait InstanceState: Clone + Debug + Send + Sync {}
 /// produce a new state, with the modifications from the block applied
 /// ([`validate_and_apply_header`](`ValidatedState::validate_and_apply_header))
 pub trait ValidatedState<TYPES: NodeType>:
-    Serialize + DeserializeOwned + Debug + Default + Hash + PartialEq + Eq + Send + Sync
+    Clone + Serialize + DeserializeOwned + Debug + Default + Hash + PartialEq + Eq + Send + Sync
```
[the remark this answers is not in the document]
Comment (shenkeyao) [reply crates/types/src/traits/states.rs]: Fixed! I missed some places when adding `Arc` so `Clone` was needed, but it's not anymore.

File: crates/types/src/event.rs [hunk 3]
```diff
@@ -22,7 +22,12 @@ pub struct Event<TYPES: NodeType> {
 }
 
 /// The chain of leafs decided on with corresponding VID info if we have it
-pub type LeafChain<TYPES> = Vec<(Leaf<TYPES>, Option<VidDisperse<TYPES>>)>;
+pub type LeafChain<TYPES> = Vec<(
+    Leaf<TYPES>,
+    Arc<<TYPES as NodeType>::ValidatedState>,
+    Option<Arc<<<TYPES as NodeType>::ValidatedState as ValidatedState<TYPES>>::Delta>>,
+    Option<VidDisperse<TYPES>>,
```
Comment (jbearer) [inline crates/types/src/event.rs:29]: This tuple is getting unwieldy, let's make a struct for it

**Review by jbearer — COMMENTED** [review]
Thanks @shenkeyao! Logic looks good I think. I had one code cleanliness comment then I think this is good

**Review by jbearer — APPROVED** [review]

**Review by imabdulbasit — APPROVED** [review]

**Review by imabdulbasit — CHANGES_REQUESTED** [review]
merge conflicts

**Comment by shenkeyao** [comment]
@imabdulbasit @jbearer Some files in `hotshot` were moved to a new `hotshot-types` repo. I'm going to fix the conflicts due to this.

**Review by imabdulbasit — APPROVED** [review]

**Pull request merged**